R/build-search-docs.R

```diff
@@ -65,3 +68,231 @@ url_node <- function(url) {
     paste0("<url><loc>", url, "</loc></url>")
   )
 }
+
+#' Build search index
+#'
+#' Build a JSON file encompassing all HTML pages, for use by the search script.
+#'
+#' @section YAML config:
+#' You can exclude some paths from the search index.
+#' Below we exclude the changelog from the search index.
+#'
+#' ```yaml
+#' search:
+#'   exclude: ['news/index.html']
+#' ```
+#' @section Local testing:
+#' Search won't work if you simply use pkgdown preview of the static files.
+#' You can use `servr::httw("docs")` instead.
+#'
+#'
+#' @inheritParams build_articles
+#' @export
+#'
+build_search <- function(pkg = ".",
+                         override = list()) {
+  pkg <- section_init(pkg, depth = 1L, override = override)
+  rule("Building search index")
+
+  paths <- fs::path_rel(
+    fs::dir_ls(pkg$dst_path, glob = "*.html", recurse = TRUE),
+    pkg$dst_path
+  )
+  paths <- paths[!paths %in% c("404.html", "articles/index.html", "reference/index.html")]
+
+  # user-defined exclusions
+  paths <- paths[!paths %in% pkg$meta$search$exclude]
+
+  if ("news/index.html" %in% paths) {
+    index <- lapply(paths[paths != "news/index.html"], file_search_index, pkg = pkg)
+    index <- unlist(index, recursive = FALSE)
+    index <- c(index, news_search_index("news/index.html", pkg = pkg))
+  } else {
+    index <- lapply(paths, file_search_index, pkg = pkg)
+    index <- unlist(index, recursive = FALSE)
+  }
+
+  index <- purrr::compact(index)
+  jsonlite::write_json(
+    index,
+    file.path(pkg$dst_path, "search.json"),
+    auto_unbox = TRUE
+  )
+}
+
+news_search_index <- function(path, pkg) {
+  html <- xml2::read_html(file.path(pkg$dst_path, path))
+
+  # Get contents minus logo
+  node <- xml2::xml_find_all(html, ".//div[contains(@class, 'contents')]")
+  xml2::xml_remove(xml2::xml_find_first(node, ".//img[contains(@class, 'pkg-logo')]"))
+  sections <- xml2::xml_find_all(node, ".//*[contains(@class, 'section')]")
+
+  # Headings (where in the page)
+  get_headings <- function(section) {
+    if (get_section_level(section) < 4) {
+      return("")
+    }
+
+    headings <- purrr::map_chr(seq(3, get_section_level(section) - 1), get_h, section = section) %>%
+      purrr::discard(function(x) x == "")
+
+    paste0(headings, collapse = " > ")
+  }
+
+  purrr::pmap(
+    list(
+      sections,
+      purrr::map_chr(sections, get_headings),
+      title = purrr::map_chr(sections, get_version)
```

Comment:
it felt easier (for reading later) to create a separate function with the logic for the changelog